core/sql/common/ComSecurityKey.cpp

```diff
@@ -73,6 +74,59 @@ NABoolean qiCheckForInvalidObject (const Int32 numInvalidationKeys,
   return found;
 }
 
+// ****************************************************************************
+// Function that builds query invalidation keys for privileges. A separate
+// invalidation key is added for each granted DML privilege. 
+// ****************************************************************************
+bool buildSecurityKeySet(PrivMgrUserPrivs *privInfo,
+                         Int64 objectUID,
+                         ComSecurityKeySet &secKeySet)
+{
+  Int32 granteeID = ComUser::getCurrentUser();
+
+  // Build security keys for object privileges
+  for ( Int32 i = FIRST_DML_PRIV; i <= LAST_DML_PRIV; i++ )
+  {
+    if ( privInfo->hasObjectPriv((PrivType)i) )
+      {
+        ComSecurityKey key(granteeID,
+                           objectUID,
+                           PrivType(i),
+                           ComSecurityKey::OBJECT_IS_OBJECT);
+        if (!key.isValid())
+          return false;
+        secKeySet.insert(key);
```

Comment:
Do we care if we return false and secKeySet is not empty (because maybe the first few keys were OK)? (I'll find out when I find this method's caller.)